ICE in insert_associated_trait_impl, at rust/typecheck/rust-typecheck-context.cc:255

<!--
File: /tmp/im/1619F7847F312C67EBA33A4C9BAC0BB6757EE339512CFDA61CA540B3D650068E.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: -frust-incomplete-and-experimental-compiler-do-not-use
pub trait Foo {}

pub struct FooImpl;

const foo_impl: () = {
    impl Foo for FooImpl {
        fn foo() {
            let () = foo_impl;
        }
    }
};

````

original:
````rust
pub trait Foo {
    type Bar;
    fn foo(bar: Self::bar);
}

pub struct FooImpl;

const foo_impl: () = {
    impl Foo for FooImpl {
        type Bar = ();

        fn foo(_bar: Self::Bar) {
            let () = foo_impl; // non-exhaustive pattern: `_` not covered reported here
        }
    }
};
````

Version information:

4db0cf74f64086e2bb2eff029c597239df8442e1

Possibly related line of code:
https://github.com/Rust-GCC/gccrs/blob/4db0cf74f64086e2bb2eff029c597239df8442e1/gcc/rust/typecheck/rust-typecheck-context.cc#L249-L261

Command:
`/home/matthias/vcs/github/gccrs/gccrs-build/gcc/crab1 -frust-incomplete-and-experimental-compiler-do-not-use`

<details><summary><strong>Program output</strong></summary>
<p>

```
/tmp/icemaker_global_tempdir.J4KdGj8QZ6tS/rustc_testrunner_tmpdir_reporting.rY4R5JiGIHX4/mvce.rs:7:9: error: method ‘foo’ is not a member of trait ‘Foo’
    6 |     impl Foo for FooImpl {
      |          ~~~
    7 |         fn foo() {
      |         ^~
/tmp/icemaker_global_tempdir.J4KdGj8QZ6tS/rustc_testrunner_tmpdir_reporting.rY4R5JiGIHX4/mvce.rs:7:9: error: method ‘foo’ is not a member of trait ‘Foo’
    6 |     impl Foo for FooImpl {
      |          ~~~
    7 |         fn foo() {
      |         ^~
crab1: internal compiler error: in insert_associated_trait_impl, at rust/typecheck/rust-typecheck-context.cc:255
0x301e249 internal_error(char const*, ...)
	../../gcc/diagnostic-global-context.cc:786
0xc22881 fancy_abort(char const*, int, char const*)
	../../gcc/diagnostics/context.cc:1787
0x9091c3 Rust::Resolver::TypeCheckContext::insert_associated_trait_impl(unsigned int, Rust::Resolver::AssociatedImplTrait&&)
	../../gcc/rust/typecheck/rust-typecheck-context.cc:255
0x103f8e2 Rust::Resolver::TypeCheckItem::validate_trait_impl_block(Rust::Resolver::TraitReference*, Rust::HIR::ImplBlock&, Rust::TyTy::BaseType*, std::vector<Rust::TyTy::SubstitutionParamMapping, std::allocator<Rust::TyTy::SubstitutionParamMapping> >&)
	../../gcc/rust/typecheck/rust-hir-type-check-item.cc:855
0x1044bcd Rust::Resolver::TypeCheckItem::visit(Rust::HIR::ImplBlock&)
	../../gcc/rust/typecheck/rust-hir-type-check-item.cc:493
0x103a985 Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
	../../gcc/rust/typecheck/rust-hir-type-check-item.cc:57
0x1075c6b ???
	../../gcc/rust/typecheck/rust-hir-type-check-stmt.h:47
0x1076143 Rust::Resolver::TypeCheckStmt::Resolve(Rust::HIR::Stmt&)
	../../gcc/rust/typecheck/rust-hir-type-check-stmt.cc:34
0x1067904 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)
	../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:614
0x1065f44 Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr&)
	../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:50
0x103a0b3 Rust::Resolver::TypeCheckItem::visit(Rust::HIR::ConstantItem&)
	../../gcc/rust/typecheck/rust-hir-type-check-item.cc:448
0x103a985 Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
	../../gcc/rust/typecheck/rust-hir-type-check-item.cc:57
0xfda37b Rust::Resolver::TypeResolution::Resolve(Rust::HIR::Crate&)
	../../gcc/rust/typecheck/rust-hir-type-check.cc:71
0xe8752c Rust::Session::compile_crate(char const*)
	../../gcc/rust/rust-session-manager.cc:682
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.

```

</p>
</details>

--- a/gcc/rust/typecheck/rust-hir-type-check.h
+++ b/gcc/rust/typecheck/rust-hir-type-check.h
@@ -235,7 +235,7 @@ public:
   void insert_trait_reference (DefId id, TraitReference &&ref);
   bool lookup_trait_reference (DefId id, TraitReference **ref);
 
-  void insert_associated_trait_impl (HirId id,
+  bool insert_associated_trait_impl (HirId id,
 				     AssociatedImplTrait &&associated);
   bool lookup_associated_trait_impl (HirId id,
 				     AssociatedImplTrait **associated);

--- a/gcc/rust/typecheck/rust-typecheck-context.cc
+++ b/gcc/rust/typecheck/rust-typecheck-context.cc
@@ -248,13 +248,17 @@ TypeCheckContext::lookup_trait_reference (DefId id, TraitReference **ref)
   return true;
 }
 
-void
+bool
 TypeCheckContext::insert_associated_trait_impl (
   HirId id, AssociatedImplTrait &&associated)
 {
-  rust_assert (associated_impl_traits.find (id)
-	       == associated_impl_traits.end ());
+  auto it = associated_impl_traits.find (id);
+  if (it != associated_impl_traits.end ())
+    {
+      return false;
+    }
   associated_impl_traits.emplace (id, std::move (associated));
+  return true;
 }
 
 bool

--- a/gcc/testsuite/rust/compile/issue-4166.rs
+++ b/gcc/testsuite/rust/compile/issue-4166.rs
@@ -1,0 +1,19 @@
+
+pub trait Foo {
+    type Bar;
+    fn foo(bar: Self::bar); // { dg-error "failed to resolve path segment using an impl Probe" }
+}
+
+pub struct FooImpl;
+
+const foo_impl: () = {
+    impl Foo for FooImpl {
+        type Bar = ();
+        fn foo(_bar: Self::Bar) { // { dg-error "method .foo. has an incompatible type|mismatched types" }
+            // This is the recursive reference that used to cause the ICE
+            let () = foo_impl; 
+        }
+    }
+};
+
+fn main() {}

